lib/mlir/Dialect/Clift/Utils/CBackend.cpp

```diff
@@ -0,0 +1,1247 @@
+  }
+
+  //===---------------------------- Statements ----------------------------===//
+
+  RecursiveCoroutine<void> emitLocalVariableDeclaration(LocalVariableOp S) {
+    // TODO: Emit variable name from the model once the model is extended to
+    //       provide this information.
+
+    rc_recur emitDeclaration(S.getResult().getType(),
+                             C.getVariableLocationDefinition(S.getSymName(),
+                                                             *CurrentFunction));
+
+    if (not S.getInitializer().empty()) {
+      Out << ' ' << '=' << ' ';
+
+      // Comma expressions in a variable initialiser must be parenthesized.
+      CurrentPrecedence = OperatorPrecedence::Comma;
+
+      rc_recur emitExpressionRegion(S.getInitializer());
+    }
+
+    Out << ';' << '\n';
+  }
+
+  RecursiveCoroutine<void> emitLabelStatement(AssignLabelOp S) {
+    Out.unindent();
+
+    Out << S.getLabelOp().getName() << ':';
+
+    // Until C23, labels cannot be placed at the end of a block.
+    if (S.getOperation() == &S->getBlock()->back())
+      Out << ' ' << ';';
+
+    Out.indent();
+
+    Out << '\n';
+
+    rc_return;
+  }
+
+  RecursiveCoroutine<void> emitExpressionStatement(ExpressionStatementOp S) {
+    rc_recur emitExpressionRegion(S.getExpression());
+    Out << ';' << '\n';
+  }
+
+  RecursiveCoroutine<void> emitGotoStatement(GoToOp S) {
+    Out << C.getKeyword(Keyword::Goto) << ' ' << S.getLabelOp().getName() << ';'
+        << '\n';
+
+    rc_return;
+  }
+
+  RecursiveCoroutine<void> emitBreakStatement() {
+    Out << C.getKeyword(Keyword::Break) << ';' << '\n';
+    rc_return;
+  }
+
+  RecursiveCoroutine<void> emitContinueStatement() {
+    Out << C.getKeyword(Keyword::Continue) << ';' << '\n';
+    rc_return;
+  }
+
+  RecursiveCoroutine<void> emitReturnStatement(ReturnOp S) {
+    Out << C.getKeyword(Keyword::Return) << ' ';
+
+    if (not S.getResult().empty())
+      rc_recur emitExpressionRegion(S.getResult());
+
+    Out << ';' << '\n';
+  }
+
+  RecursiveCoroutine<void> emitBlockStatement(BlockStatementOp S) {
+    Out << '{' << '\n';
+    enter(), rc_recur emitStatementRegion(S.getBody());
+    Out << '}' << '\n';
+  }
+
+  RecursiveCoroutine<void> emitIfStatement(IfOp S) {
+    // Nested if-else-if chains are printed out in a loop to avoid introducing
+    // extra indentation for each else-if.
+    while (true) {
+      Out << C.getKeyword(Keyword::If) << ' ' << '(';
+      rc_recur emitExpressionRegion(S.getCondition());
+      Out << ')';
+
+      bool Braces = rc_recur emitImplicitBlockStatement(S.getThen());
+
+      if (S.getElse().empty())
+        break;
+
+      if (Braces)
+        Out << ' ';
+      Out << C.getKeyword(Keyword::Else);
+
+      if (auto ElseIf = getOnlyOperation<IfOp>(S.getElse())) {
+        S = ElseIf;
+        Out << ' ';
+      } else {
+        if (rc_recur emitImplicitBlockStatement(S.getElse()))
+          Out << '\n';
+        break;
+      }
+    }
```

Comment:
I think you're right. That's a good catch. I wonder if the brace elision logic should instead just check for a single expression statement? Instead of any single statement? I'm not sure something like this is desirable anyway:
```C
if (condition)
  while (other_condition) {
    ...
  }
```

Reply:
```C
void fun_0x40001001(void) {
  if (0)
    1;
  else if (2)
    3;
  else
    4;
  if (5) {
    6;
    7;
  } else if (8) {
    9;
    10;
  } else {
    11;
    12;
  }
  if (13) {
    if (14)
      15;
  } else
    16;
}
```
I changed it to produce this output. Does that seem reasonable? The elided braces on the `else` look a little weird admittedly. Do we want to specify more complicated rules for brace elision? E.g. only elide braces if they can be elided for all branches?

Reply:
It now outputs this:
```C
void fun_0x40001001(void) {
  if (0)
    1;
  else if (2)
    3;
  else
    4;
  if (5) {
    6;
    7;
  } else if (8) {
    9;
    10;
  } else {
    11;
    12;
  }
  if (13) {
    if (14)
      15;
  } else {
    16;
  }
}
```

The code is a bit more complicated, as it now requires two passes over the if-else chain.

Reply:
Ok, the functionality is now what I expect.

I think the code can be improved a bit, to make it a little easier to follow even if it's more complex.

I'd outline the first loop, the one where you compute the `EmitBlocks` boolean condition, in a separate function (`shouldEmitBraces` or something), to be implemented with `RecursiveCoroutine`.

Reply:
Deciding whether the braces can be elided is now moved into a separate function, but not using `RecursiveCoroutine.`